src/aspire/utils/rotation.py

```diff
@@ -0,0 +1,280 @@
+"""
+Define a Rotation Class for customized rotation operations used by ASPIRE.
+"""
+
+import copy
+
+import numpy as np
+from numpy.linalg import norm
+from scipy.linalg import svd
+from scipy.spatial.transform import Rotation as sp_rot
+
+from aspire.utils import ensure
+from aspire.utils.random import Random
+
+
+class Rotation:
+    def __init__(self, num_rots, seed=None, seq="ZYZ", angles=None, dtype=np.float32):
+        """
+        Initialize a Rotation object
+
+        :param num_rots: Total number of rotation sets
+        :param seed: Seed for initializing random rotations.
+            If None, the random generator is not re-seeded.
+        :param seq:  Sequence of order applying Euler angles
+        :param angles: Euler angles (in degrees) to generate rotation matrices.
+            If None, uniformly distributed angles will be generated.
+        :param dtype: data type for angles and rotation matrices
+        """
+        self.rot_seq = seq
+        self.dtype = np.dtype(dtype)
+        if angles:
+            ensure(
+                num_rots == angles.shape[0],
+                "Number of rotation matrices should equal to "
+                "number of sets of Euler angles.",
+            )
+            self.angles = angles.astype(self.dtype)
+        else:
+            self.angles = self._uniform_random_angles(num_rots, seed=seed)
+        self.num_rots = num_rots
+
+        self.data = self.rot_matrices
+        self.shape = (num_rots, 3, 3)
+
+    def __str__(self):
+        """
+        String representation.
+        """
+        return (
+            f"Rotation object with matrices[{self.num_rots}, 3, 3] of {self.dtype} type"
+        )
+
+    def __len__(self):
+        return self.num_rots
+
+    def __getitem__(self, item):
+        return self.data[item]
+
+    def __setitem__(self, key, value):
+        self.data[key] = value
+
+    def __mul__(self, other):
+        if isinstance(other, Rotation):
+            output = self.data @ other.data
+        else:
+            output = self.data * other
+        return output
+
+    def _uniform_random_angles(self, n, seed=None):
+        """
+        Generate random 3D rotation angles in degrees
+
+        :param n: The number of rotation angles to generate
+        :param seed: Random integer seed to use. If None,
+            the current random state is used.
+        :return: A n-by-3 ndarray of rotation angles in degrees
+        """
+        # Generate random rotation angles, in radians
+        angles = np.zeros((n, 3), dtype=self.dtype)
+        with Random(seed):
+            angles = np.column_stack(
+                (
+                    np.random.random(n) * 2 * np.pi,
+                    np.arccos(2 * np.random.random(n) - 1),
+                    np.random.random(n) * 2 * np.pi,
+                )
+            )
+        # Return random rotation angles in degrees
+        return np.rad2deg(angles).astype(self.dtype)
```

Comment:
Let's stick to radians as much as possible.

Reply:
Changed to use radians in commit 0c8f0b8.